### PR TITLE
Simplify simulator

### PR DESCRIPTION
**Pull request by andlaus**
following on the tracks of #385, this applies the same tricks to the simulator class. The main benefiter is opm-polymer, but this is a separate PR.

as far as I can tell, there are no performance regressions for SPE-1 and SPE-9 from opm-data.

### CODE REVIEW

File: opm/autodiff/NewtonSolver.hpp [hunk 1]
```diff
@@ -66,7 +66,7 @@ namespace Opm {
         /// \param[in]      param   parameters controlling nonlinear Newton process
         /// \param[in, out] model   physical simulation model
         explicit NewtonSolver(const SolverParameters& param,
-                              PhysicalModel& model);
+                              std::shared_ptr<PhysicalModel> model);
```
Comment (atgeirr) [inline opm/autodiff/NewtonSolver.hpp]: Why did you change this to use a shared_ptr?

Comment (bska) [inline opm/autodiff/NewtonSolver.hpp]: More to the point, why is there a `shared_ptr<>` in the interface at all given the inferred semantics as detailed in Herb Sutter's [GOTW 91](http://herbsutter.com/2013/06/05/gotw-91-solution-smart-pointer-parameters/)?

File: opm/autodiff/SimulatorBase_impl.hpp [hunk 2]
```diff
@@ -18,167 +18,21 @@
...
-                                               BlackoilPropsAdInterface& props,
-                                               const RockCompressibility* rock_comp_props,
-                                               NewtonIterationBlackoilInterface& linsolver,
-                                               const double* gravity,
-                                               const bool has_disgas,
-                                               const bool has_vapoil,
-                                               std::shared_ptr<EclipseState> eclipse_state,
-                                               BlackoilOutputWriter& output_writer,
-                                               const std::vector<double>& threshold_pressures_by_face)
+    template<class Implementation>
```
Comment (atgeirr) [inline opm/autodiff/SimulatorBase_impl.hpp]: So it is you who have been putting these no-space-between-template-and-left-bracket lines in the code :-)

I really think it looks a lot better with a space after `template`.

**Comment by atgeirr** [comment]
I think that this should go in (after revision), but I'll have to continue my review later.

File: opm/autodiff/NewtonSolver.hpp (hunk 1 above)
Comment (andlaus) [inline opm/autodiff/NewtonSolver.hpp]: @bska: I don't really understand your question. sorry for my stupidity.
@atgeirr: the shared pointer is necessary to allow the model and the solver to be allocated in a separate method. currently it only worked because the solver and the model live in the same scope (i.e., the `run()` method), thus the model got deallocated after the solver.

File: opm/autodiff/SimulatorBase_impl.hpp (hunk 2 above)
Comment (andlaus) [inline opm/autodiff/SimulatorBase_impl.hpp]: IMO, it looks better without the empty lines, but actually I don't really care. changed.

File: opm/autodiff/NewtonSolver.hpp (hunk 1 above)
Comment (bska) [inline opm/autodiff/NewtonSolver.hpp]: If you use a smart pointer as a parameter type, you're implicitly stating that you want to take part in the lifetime management of the underlying object.  Unless you want to do that--and there are certainly cases where you do want to do that--you should accept the parameter as a reference (typically) or a pointer and not care how the caller chooses to manage the object in question.

See also Sutter's [Back to Basics](https://www.youtube.com/watch?v=xnqTKD8uD64) talk from cppcon 2014 (1 hour, 40 minutes duration) about the same and related subjects.

Comment (andlaus) [inline opm/autodiff/NewtonSolver.hpp]: > If you use a smart pointer as a parameter type, you're implicitly stating that you want to take part in the lifetime management of the underlying object. Unless you want to do that--and there are certainly cases where you do want to do that--you should accept the parameter as a reference (typically) or a pointer and not care how the caller chooses to manage the object in question.

object lifecycle management was the whole reason for this change, see my answer to @atgeirr above. In this context, be aware that this constructor is only called once per report step, i.e., the performance implications mentioned by Sutter are not even measurable. (if you want to have some serious fun with `shared_ptr` performance, you should probably have a look at opm-parser.)

**Comment by joakim-hove** [comment]
> (if you want to have some serious fun with shared_ptr performance, you should probably have a look at opm-parser.)

What - don't mess with my shared pointers :angry:. Seriously I read the Sutter article with interest and intend to watch talk @bska mentioned.

File: opm/autodiff/NewtonSolver.hpp (hunk 1 above)
Comment (bska) [inline opm/autodiff/NewtonSolver.hpp]: > object lifecycle management was the whole reason for this change

Well, in that case you're probably fine although a comment to that effect would not og amiss.

> the performance implications mentioned by Sutter are not even measurable

I didn't care too much about that (on this occasion).  I was more concerned about the correctness implications of a smart pointer parameter.

Comment (andlaus) [inline opm/autodiff/NewtonSolver.hpp]: > a comment to that effect would not go amiss.

okay. I'll add one tomorrow.

> I was more concerned about the correctness implications of a smart pointer parameter.

I thought that with `shared_ptr` everything is golden as long as there are no cycles. If there are, they can still be used, but weak references need to be inserted to break them. maybe I'm wrong, though...

Comment (atgeirr) [inline opm/autodiff/NewtonSolver.hpp]: The correctness argument is that you should use a reference unless you will actually store the pointer and thereby participate in lifecycle mangement. Since you actually store it, using it as a function argument is ok actually, so my question should have been: why have you changed the *data member* to be a shared_ptr? I do not see the need upfront, did you change something in the way the classes are used that makes you require this?

File: opm/autodiff/SimulatorBase_impl.hpp (hunk 2 above)
Comment (atgeirr) [inline opm/autodiff/SimulatorBase_impl.hpp]: I did not want an extra line, but a space like this: `template <class Foo>` instead of `template<class Foo>`. For that matter I do not see a change in the diff?

Comment (atgeirr) [inline opm/autodiff/SimulatorBase_impl.hpp]: Now I see what you changed, sorry for not noticing: yes I think an extra line is an improvement where you put it in your latest commit, but that was not what I asked for (as explained above)...

**Comment by atgeirr** [comment]
> > (if you want to have some serious fun with shared_ptr performance, you should probably have a look at opm-parser.)

> What - don't mess with my shared pointers . Seriously I read the Sutter article with interest and intend to watch talk @bska mentioned.

I do think we need not fear shared_ptr performance: the parser seems quite snappy even using them all over the place. Still, we are only talking about a few thousand of them, not a million. I know another project that used shared_ptr for each of several millions of particles in a particle sim: it worked, but they did have to be careful not to copy the shared_ptrs for example.

I am still not quite sold that it was the *best* way to build the opm-parser, but it proved a long time ago to be a *more than good enough* way.

File: opm/autodiff/SimulatorBase.hpp [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+        /// modify the reservoir and well states.
+        /// \param[in,out] timer       governs the requested reporting timesteps
+        /// \param[in,out] state       state of reservoir: pressure, fluxes
+        /// \param[in,out] well_state  state of wells: bhp, perforation rates
+        /// \return                    simulation report, with timing data
+        SimulatorReport run(SimulatorTimer& timer,
+                            ReservoirState& state);
+
+    protected:
+        Implementation& asImp_() { return *static_cast<Implementation*>(this); }
```
Comment (atgeirr) [inline opm/autodiff/SimulatorBase.hpp]: I already uses asImpl() for the other classes, and suggest that you do, too (also I do not like the private methods ending in underscore, without really understanding why).

File: opm/autodiff/SimulatorBase.hpp [hunk 4]
```diff
@@ -0,0 +1,195 @@
+/*
+  Copyright 2013 SINTEF ICT, Applied Mathematics.
```
Comment (atgeirr) [inline opm/autodiff/SimulatorBase.hpp]: It seems to me that you have made sufficient changes to warrant adding your own copyright to this. I assume everybody treats this in a reasonable and sensible manner...

File: opm/autodiff/SimulatorBase_impl.hpp [hunk 5]
```diff
@@ -475,13 +318,47 @@ namespace Opm
...
+    template <class Implementation>
+    void SimulatorBase<Implementation>::handleAdditionalWellInflow(SimulatorTimer& timer,
+                                                                   WellsManager& wells_manager,
+                                                                   WellState& well_state,
+                                                                   const Wells* wells)
+    { }
+
+    template <class Implementation>
+    auto SimulatorBase<Implementation>::createSolver(const Wells* wells)
+        -> std::shared_ptr<Solver>
```
Comment (atgeirr) [inline opm/autodiff/SimulatorBase_impl.hpp]: Is this syntax supported by gcc 4.4? If so I am fine with its use.

**Comment by atgeirr** [comment]
I have now read through this, and I see why you made the NewtonSolver take a shared_ptr. I still think that change should be reversed, however. The construction of the NewtonSolver can be changed to take place in run() rather than in createSolver(), and a method createModel() can be used instead (it would be just like the first part of the current createSolver()).

If we decide in the future that we actually need to create the solver in a separate function, then I may reconsider, but I do not see the need now. Also, if it comes to that a unique_ptr (explicit object sink to transfer ownership to the NewtonSolver) is probably better, since I can only see trouble from actually sharing a model between solver instances.

**Comment by atgeirr** [comment]
I'll add that now that I have looked at the polymer module I see that you need the distinct createSolver() method there, for the case that does not use the NewtonSolver. So I have changed my mind, since we actually *do* already have a use-case. I still think that all shared_ptr usage in these cases should be replaced with unique_ptr.

**Comment by andlaus** [comment]
> I still think that all shared_ptr usage in these cases should be replaced with unique_ptr.

`unique_ptr` assumes -- as its name suggests -- that there is only one pointer to a given object. This implies that it cannot be used for parameter passing and -- more importantly -- the `model` object will be deleted at the end of the `createSolver()` method. if you're really dogmatic about this, one can either create the model inside the solver (provided that they can all be created the same way) or use a raw pointer to the model from which the `unique_ptr` gets assigned. I think both options come with quite a few additional implicit assumptions which would *definitely* need *more* explaination  than just using `shared_ptr`; in particular considering the fact that the runtime overhead of `shared_ptr` is essentially zero in this case.

tl:dr: I'll change it for you, but only after you explicitly state "I'm stubbornly dogmatic about this!"

File: opm/autodiff/SimulatorBase_impl.hpp (hunk 5 above)
Comment (andlaus) [inline opm/autodiff/SimulatorBase_impl.hpp]: yep, it's supported. in 4.4 eWoms uses this in conjunction with `decltype` to implement perfect forwarding.

File: opm/autodiff/SimulatorBase.hpp (hunk 4 above)
Comment (andlaus) [inline opm/autodiff/SimulatorBase.hpp]: as far as I see it, these headers do not have any legal relevance anyway, but I'll add myself.

File: opm/autodiff/SimulatorBase.hpp (hunk 3 above)
Comment (andlaus) [inline opm/autodiff/SimulatorBase.hpp]: > (also I do not like the private methods ending in underscore, without really understanding why).

the reason is to immediately see which members/attributes are public and which are not when reading a piece of code. since this convention is used in some places in opm, but not in others, I changed the underscore by an 'l', as you requested.

**Comment by atgeirr** [comment]
> unique_ptr assumes -- as its name suggests -- that there is only one pointer to a given object. This implies that it cannot be used for parameter passing and -- more importantly -- the model object will be deleted at the end of the createSolver() method

Yes, there will be only one valid pointer at a time. It can be used for parameter passing, it just implies that you then lose the ability to use it on the outside. Passing a unique_ptr is a well-known and accepted pattern for passing ownership. So passing a unique_ptr to a constructor implies that the object constructed will now own the unique_ptr. This of course requires move semantics, but that is present in gcc 4.4 so it should be all right (unique_ptrs are moveable). The model object will not be deleted at the end of createSolver() since a unique_ptr which has been moved from no longer owns or references anything.

**Comment by andlaus** [comment]
> It can be used for parameter passing

really? g++ 4.8 bails out for the following program:

```c++
#include <memory>
#include <iostream>

void foo(std::unique_ptr<int> a)
{ std::cout << "i am foo() and get " << *a << "\n"; }

int main()
{
    std::unique_ptr<int> bar(new int(123));
    foo(bar);
    std::cout << "i am main() and get " << *bar << "\n";
    return 0;
};
```

the error I get is:

```
and@heuristix:~/tmp > g++ -std=c++11 foo.cc && ./a.out 
foo.cc: In function ‘int main()’:
foo.cc:10:12: error: use of deleted function ‘std::unique_ptr<_Tp, _Dp>::unique_ptr(const std::unique_ptr<_Tp, _Dp>&) [with _Tp = int; _Dp = std::default_delete<int>]’
     foo(bar);
            ^
In file included from /usr/include/c++/4.8/memory:81:0,
                 from foo.cc:1:
/usr/include/c++/4.8/bits/unique_ptr.h:273:7: error: declared here
       unique_ptr(const unique_ptr&) = delete;
       ^
foo.cc:4:6: error:   initializing argument 1 of ‘void foo(std::unique_ptr<int>)’
 void foo(std::unique_ptr<int> a)
      ^
```

i.e. the copy constructor has been explicitly deleted. (it does not use the move constructor, but that's the same situation as in `createSolver()`).

**Comment by bska** [comment]
> the copy constructor has been explicitly deleted

Yes.  You need to call
```C++
foo(std::move(bar));
```
(and `#include <algorithm>`) which, although somewhat cumbersome, is at least very explicit about transferring ownership to `foo()`.

**Comment by bska** [comment]
I didn't notice until now, but this change broke `make install`.  The renaming of `SimulatorFullyImplicitBlackoil_impl.hpp` to `SimulatorBase_impl.hpp` must be recorded in `CMakeLists_files.txt`.
